README
SAP OCI
=======

Ruby library and Rails plugin for punchout via SAP OCI protocol.


Copyright (c) 2011 Oliver Eilhard, released under the MIT license


CHANGELOG.md
# 0.2.0, released 2015-07-22

* Update gems
* Require nokogiri ~> 1.6
* Require Ruby ~> 2.2
* Ignore unknown properties in OCI item, e.g. NEW_ITEM-TAX_RATE

# 0.1.9, released 2014-03-14

* Added dependency on Nokogiri
* Added development dependency on Rake (for REE 1.8.7)

# 0.1.8, released 2012-05-06

* Add test for HTML-escaping fields, e.g. longtext
* Remove rake dependency
* Tested under REE 1.8.7 and 1.9.3-p125

# 0.1.7, released 2012-02-08

* Replace comma by dot in numeric fields

# 0.1.6, released 2011-09-22

* Wrong tests for longtext, but they still succeed
* Longtext not parsed in HTML case

# 0.1.5, released 2011-09-06

* 1 is returned if PRICEUNIT is missing, empty, or whitespace

# 0.1.4, release 2011-07-14

* Add various before_type_cast methods

# 0.1.3, release 2011-07-14

* Ignore whitespace on numeric fields

# 0.1.2, release 2011-07-05

* Use Nokogiri 1.5.0

# 0.1.1, release 2011-04-30

* Removed version from sources and put it into gemspec only

# 0.1.0, released 2011-04-29

* Added a gemspec
* Initial release on GitHub



test/test_parse.rb
# encoding: utf-8

require 'sapoci'
require 'sapoci/document'
require 'test/unit'

class TestParse < Test::Unit::TestCase
  include SAPOCI

  def test_parse_html
    Document.from_html(valid_single_html) do |doc|
      count = 0
      doc.items.each do |item|
        assert_equal "Description 1", item.description
        count += 1
      end
      assert_equal 1, count
    end
  end

  def test_parse_long_html
    Document.from_html(valid_long_html) do |doc|
      count = 0
      assert_equal 2, doc.items.count
      assert_equal "Apple MacBook Air 11\"", doc.items[0].description
      assert_equal "Ein tolles Notebook von Apple.", doc.items[0].longtext
      assert_equal "Apple iMac 27\"", doc.items[1].description
      assert_equal "Der elegante Desktop-Rechner von Apple.", doc.items[1].longtext
    end
  end

  def test_parse_real_world
    file = File.expand_path(File.dirname(__FILE__) + "/files/real_world.html")
    Document.from_html(IO.read(file)) do |doc|
      count = 0
      doc.items.each do |item|
        assert !item.description.blank?
        count += 1
      end
      assert_equal 10, count
    end
  end

  def test_parse_params
    Document.from_params(valid_single_params) do |doc|
      count = 0
      doc.items.each do |item|
        assert_equal "Visitenkarten", item.description
        assert_equal "Standard Visitenkarte deutsch 200 St. ", item.longtext
        assert_equal 16.18, item.price
        count += 1
      end
      assert_equal 1, count
    end
  end

  def test_ignore_unknown_fields
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-TAX_RATE"=>{"1"=>"0.19"},  # <= no such property in OCI
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>"100 "},  # <= watch the whitespace
      "NEW_ITEM-QUANTITY"=>{"1"=>" 12 "},  # <= watch the whitespace
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 100.00, item.priceunit
      assert_equal 12, item.quantity
    end
  end

  def test_ignore_whitespace_on_numeric_fields
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-PRICE"=>{"1"=>" 780.00"},  # <= watch the whitespace
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>"100 "},  # <= watch the whitespace
      "NEW_ITEM-QUANTITY"=>{"1"=>" 12 "},  # <= watch the whitespace
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 780.00, item.price
      assert_equal 100.00, item.priceunit
      assert_equal 12, item.quantity
    end
  end

  def test_parse_numeric_fields_with_comma
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-PRICE"=>{"1"=>" 780,12"},  # <= watch the whitespace and the comma
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 780.12, item.price
    end
  end

  def test_return_1_if_priceunit_is_missing
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-PRICE"=>{"1"=>"780.00"},
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 1, item.priceunit
    end
  end

  def test_return_1_if_priceunit_is_empty
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-PRICE"=>{"1"=>"780.00"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>""},  # <= watch the blanks
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 1, item.priceunit
    end
  end

  def test_return_1_if_priceunit_is_all_whitespace
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-PRICE"=>{"1"=>"780.00"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>" "},  # <= watch the whitespace
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 1, item.priceunit
    end
  end

  def test_before_type_cast_methods
    params = {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-PRICE"=>{"1"=>" 780.00"},  # <= watch the whitespace
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>"100 "},  # <= watch the whitespace
      "NEW_ITEM-QUANTITY"=>{"1"=>" 12 "},  # <= watch the whitespace
    }
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert item = doc.items.first
      assert_equal 780.00, item.price
      assert_equal " 780.00", item.price_before_type_cast
      assert_equal 100.00, item.priceunit
      assert_equal "100 ", item.priceunit_before_type_cast
      assert_equal 12, item.quantity
      assert_equal " 12 ", item.quantity_before_type_cast
    end
  end

  def test_parse_non_array_longtext
    # INVALID: NEW_ITEM-LONGTEXT_1:132  (No square brackets!)
    params = valid_single_params
    params["NEW_ITEM-LONGTEXT_1:132"] = "Standard Visitenkarte deutsch 200 St. "
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert_equal "Standard Visitenkarte deutsch 200 St. ", doc.items[0].longtext
    end
  end

  def test_parse_array_longtext
    # INVALID: NEW_ITEM-LONGTEXT_1:132[]
    params = valid_single_params
    params["NEW_ITEM-LONGTEXT_1:132"] = ["Standard Visitenkarte deutsch 200 St. "]
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert_equal "Standard Visitenkarte deutsch 200 St. ", doc.items[0].longtext
    end
  end

  def test_parse_indexed_longtext
    # INVALID: NEW_ITEM-LONGTEXT_1:132[1]  (Index ist falsch!)
    params = valid_single_params
    params["NEW_ITEM-LONGTEXT_1:132"] = {"1"=>"Standard Visitenkarte deutsch 200 St. "}
    Document.from_params(params) do |doc|
      assert_equal 1, doc.items.size
      assert_equal "Standard Visitenkarte deutsch 200 St. ", doc.items[0].longtext
    end
  end

  def test_parse_order
    Document.from_html(valid_multiple_html) do |doc|
      assert_equal 3, doc.items.size
      assert_equal 0, doc.items[0].index
      assert_equal "Description 1", doc.items[0].description
      assert_equal 1, doc.items[1].index
      assert_equal "Description 2", doc.items[1].description
      assert_equal 2, doc.items[2].index
      assert_equal "Description 3", doc.items[2].description
    end
  end

  # Will parse data according to POSITION specified by input,
  # regardless of holes.
  def test_parse_invalid_order
    html = <<EOF
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[1]' value='Description 1'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[3]' value='Description 3'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[4]' value='Description 4'>
EOF
    Document.from_html(html) do |doc|
      assert_equal 3, doc.items.size
      assert_equal 0, doc.items[0].index
      assert_equal "Description 1", doc.items[0].description
      assert_equal 2, doc.items[1].index
      assert_equal "Description 3", doc.items[1].description
      assert_equal 3, doc.items[2].index
      assert_equal "Description 4", doc.items[2].description
    end
  end

  def test_parse_quotes_from_html
    Document.from_html(valid_html_with_quotes) do |doc|
      count = 0
      doc.items.each do |item|
        assert_equal "19\" Rack", item.description
        count += 1
      end
      assert_equal 1, count
    end
  end

  def test_strip_quotes_for_invalid_input
    # Notice the quote in the value attributes
    html = '<input type="hidden" name="NEW_ITEM-DESCRIPTION[1]" value="19" Rack">'
    Document.from_html(html) do |doc|
      assert_equal 1, doc.items.count
      assert_equal "19", doc.items[0].description
    end
  end

private

  def valid_single_html
<<EOF
<html>
  <head>
    <title>OCI Data</title>
  </head>
  <body>
    <form method='POST' action='http://punchout.local/punchback'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[1]' value='Description 1'>
    </form>
  </body>
</html>
EOF
  end

  def valid_multiple_html
<<EOF
<html>
  <head>
    <title>OCI Data</title>
  </head>
  <body>
    <form method='POST' action='http://punchout.local/punchback'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[2]' value='Description 2'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[3]' value='Description 3'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[1]' value='Description 1'>
    </form>
  </body>
</html>
EOF
  end

  def valid_long_html
<<EOF
<html>
  <head><title>Search1</title></head>
  <body>
    <form method="POST" action="http://return.to/me">
      <input type="hidden" name="NEW_ITEM-DESCRIPTION[1]" value="Apple MacBook Air 11&quot;">
      <input type="hidden" name="NEW_ITEM-QUANTITY[1]" value="1.00">
      <input type="hidden" name="NEW_ITEM-UNIT[1]" value="PCE">
      <input type="hidden" name="NEW_ITEM-PRICE[1]" value="999.90">
      <input type="hidden" name="NEW_ITEM-CURRENCY[1]" value="EUR">
      <input type="hidden" name="NEW_ITEM-PRICEUNIT[1]" value="1">
      <input type="hidden" name="NEW_ITEM-LEADTIME[1]" value="7">
      <input type="hidden" name="NEW_ITEM-VENDOR[1]" value="Apple">
      <input type="hidden" name="NEW_ITEM-VENDORMAT[1]" value="MBA11">
      <input type="hidden" name="NEW_ITEM-MATGROUP[1]" value="NOTEBOOK">
      <input type="hidden" name="NEW_ITEM-LONGTEXT_1:132[]" value="Ein tolles Notebook von Apple.">

      <input type="hidden" name="NEW_ITEM-DESCRIPTION[2]" value="Apple iMac 27&quot;">
      <input type="hidden" name="NEW_ITEM-QUANTITY[2]" value="2.00">
      <input type="hidden" name="NEW_ITEM-UNIT[2]" value="PCE">
      <input type="hidden" name="NEW_ITEM-PRICE[2]" value="1799.00">
      <input type="hidden" name="NEW_ITEM-CURRENCY[2]" value="EUR">
      <input type="hidden" name="NEW_ITEM-PRICEUNIT[2]" value="1">
      <input type="hidden" name="NEW_ITEM-LEADTIME[2]" value="7">
      <input type="hidden" name="NEW_ITEM-VENDOR[2]" value="Apple">
      <input type="hidden" name="NEW_ITEM-VENDORMAT[2]" value="IMAC27">
      <input type="hidden" name="NEW_ITEM-MATGROUP[2]" value="DESKTOP">
      <input type="hidden" name="NEW_ITEM-LONGTEXT_2:132[]" value="Der elegante Desktop-Rechner von Apple.">
    </form>
  </body>
</html>
EOF
  end

  def valid_html_with_quotes
<<EOF
<html>
  <head>
    <title>OCI Data</title>
  </head>
  <body>
    <form method='POST' action='http://punchout.local/punchback'>
      <input type='hidden' name='NEW_ITEM-DESCRIPTION[1]' value='19" Rack'>
    </form>
  </body>
</html>
EOF
  end

  def valid_single_params
    {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
      "NEW_ITEM-CUST_FIELD1"=>{"1"=>""},
      "NEW_ITEM-CUST_FIELD2"=>{"1"=>""},
      "NEW_ITEM-CUST_FIELD3"=>{"1"=>""},
      "NEW_ITEM-LEADTIME"=>{"1"=>"10"},
      "NEW_ITEM-MANUFACTCODE"=>{"1"=>""},
      "NEW_ITEM-SERVICE"=>{"1"=>""},
      "NEW_ITEM-EXT_PRODUCT_ID"=>{"1"=>""},
      "NEW_ITEM-CUST_FIELD4"=>{"1"=>""},
      "NEW_ITEM-CONTRACT_ITEM"=>{"1"=>""},
      "NEW_ITEM-EXT_QUOTE_ID"=>{"1"=>""},
      "NEW_ITEM-LONGTEXT_1:132"=>["Standard Visitenkarte deutsch 200 St. "],
      "NEW_ITEM-CUST_FIELD5"=>{"1"=>""},
      "NEW_ITEM-QUANTITY"=>{"1"=>"1"},
      "NEW_ITEM-CONTRACT"=>{"1"=>""},
      "NEW_ITEM-PRICE"=>{"1"=>"16.18"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>""},
      "NEW_ITEM-VENDOR"=>{"1"=>"9999"},
      "NEW_ITEM-VENDORMAT"=>{"1"=>"1517"},
      "NEW_ITEM-MATGROUP"=>{"1"=>"24-26-09-01"},
      "NEW_ITEM-MANUFACTMAT"=>{"1"=>""},
      "NEW_ITEM-EXT_QUOTE_ITEM"=>{"1"=>""},
      "NEW_ITEM-MATNR"=>{"1"=>""},
      "NEW_ITEM-UNIT"=>{"1"=>"EA"}
    }
  end

end


test/test_output.rb
# encoding: utf-8

require 'sapoci'
require 'sapoci/document'
require 'test/unit'

class TestOutput < Test::Unit::TestCase
  include SAPOCI

  def test_edge_cases
    assert_equal "", Document.from_params(nil).to_html
    assert_equal "", Document.from_params({}).to_html
    assert_equal "", Document.from_params({"TEST" => "VALUE"}).to_html
  end

  def test_simple_output
    params = { "NEW_ITEM-DESCRIPTION"=>{"1"=>"Description 1"} }
    Document.from_params(params) do |doc|
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[1\]" value="Description 1"/, doc.to_html
    end
  end

  def test_escaped_output
    params = { "NEW_ITEM-DESCRIPTION"=>{"1"=>"Apple iMac\""} }
    Document.from_params(params) do |doc|
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[1\]" value="Apple iMac&quot;"/, doc.to_html
    end
  end

  def test_complex_output
    Document.from_params(valid_params_complex) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[1\]" value="Visitenkarten"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-CURRENCY\[1\]" value="EUR"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-PRICE\[1\]" value="00000000016\.180"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-LONGTEXT_1:132\[\]" value="Standard Visitenkarte deutsch 200 St. "/, html
    end
  end

  def test_output_non_array_longtext
    # INVALID: NEW_ITEM-LONGTEXT_1:132  (No square brackets!)
    params = { "NEW_ITEM-LONGTEXT_1:132" => "Standard Visitenkarte deutsch 200 St. " }
    Document.from_params(params) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-LONGTEXT_1:132\[\]" value="Standard Visitenkarte deutsch 200 St. "/, html
    end
  end

  def test_output_indexed_longtext
    # INVALID: NEW_ITEM-LONGTEXT_1:132[1]  (Index ist falsch!)
    params = { "NEW_ITEM-LONGTEXT_1:132"=>{"1"=>"Standard Visitenkarte deutsch 200 St. "} }
    Document.from_params(params) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-LONGTEXT_1:132\[\]" value="Standard Visitenkarte deutsch 200 St. "/, html
    end
  end

  def test_output_of_priceunit
    # TEST: NEW_ITEM-PRICEUNIT[1] == "1"
    params = { "NEW_ITEM-PRICEUNIT"=>{"1"=>"1"} }
    Document.from_params(params) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-PRICEUNIT\[1\]" value="1"/, html
    end
  end

  def test_output_correct_index
    params = {
      "NEW_ITEM-DESCRIPTION"=>{
        "1"=>"PRITT COMPACT KORREKTURROLLER 4,2MMX8,5M",
        "2"=>"BX100 SIGEL DP461 BRIEFPAP.IMPRESSIONS",
        "3"=>"BX250 IDEM 92100910721 SD-PAP. A4 WSGE",
        "4"=>"BX150 HP CG965A FARBLASERPAPIER A4 150G"
      },
      "NEW_ITEM-LONGTEXT_2:132"=>["Briefpapier Sigel DP461, Impressions, 100 Blatt"],
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR", "2"=>"EUR", "3"=>"EUR", "4"=>"USD"},
      "NEW_ITEM-LEADTIME"=>{"1"=>"1", "2"=>"1", "3"=>"1", "4"=>"1"
      },
      "NEW_ITEM-LONGTEXT_3:132"=>["Durchschreibepapapier Idem 2fach, weiß/gelb, 250 Blatt"],
      "NEW_ITEM-LONGTEXT_1:132"=>["Korrekturroller Pritt Compact Länge 8.5m Breite 4.2mm für Gedrucktes"],
      "NEW_ITEM-LONGTEXT_4:132"=>["Fotopapier HP CG965A beidseitig beschichtet A4 hochglanz 150g/qm 150 Blatt"],
      "NEW_ITEM-EXT_PRODUCT_ID"=>{
        "1"=>"124474#ST",
        "2"=>"4938724#ST",
        "3"=>"4678213#ST",
        "4"=>"4567185#ST"
      },
      "NEW_ITEM-MATGROUP"=>{"1"=>"24110202", "2"=>"24110534", "3"=>"24110534", "4"=>"24140301"},
      "NEW_ITEM-PRICE"=>{"1"=>"1.440", "2"=>"8.470", "3"=>"18.400", "4"=>"25.740"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>"1", "2"=>"1", "3"=>"1", "4"=>"1"},
      "NEW_ITEM-VENDORMAT"=>{"1"=>"124474", "2"=>"4938724", "3"=>"4678213", "4"=>"4567185"},
      "NEW_ITEM-QUANTITY"=>{"1"=>"1.000", "2"=>"1.000", "3"=>"1.000", "4"=>"1.000"},
      "NEW_ITEM-UNIT"=>{"1"=>"PCE", "2"=>"PCE", "3"=>"PCE", "4"=>"PCE"}
    }
    Document.from_params(params) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[1\]" value="PRITT COMPACT KORREKTURROLLER 4,2MMX8,5M"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-CURRENCY\[1\]" value="EUR"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-PRICE\[1\]" value="00000000001\.440"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[2\]" value="BX100 SIGEL DP461 BRIEFPAP.IMPRESSIONS"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-CURRENCY\[2\]" value="EUR"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-PRICE\[2\]" value="00000000008\.470"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[3\]" value="BX250 IDEM 92100910721 SD-PAP. A4 WSGE"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-CURRENCY\[3\]" value="EUR"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-PRICE\[3\]" value="00000000018\.400"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-DESCRIPTION\[4\]" value="BX150 HP CG965A FARBLASERPAPIER A4 150G"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-CURRENCY\[4\]" value="USD"/, html
      assert_match /<input type="hidden" name="NEW_ITEM-PRICE\[4\]" value="00000000025\.740"/, html
    end
  end

  def test_correct_escaping
    params = { "NEW_ITEM-LONGTEXT_1:132"=>["<script>alert('Hello');</script>"] }
    Document.from_params(params) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-LONGTEXT_1:132\[\]" value="&lt;script&gt;alert\(&#39;Hello&#39;\);&lt;\/script&gt;" \/>/, html
    end

    params = { "NEW_ITEM-LONGTEXT_1:132"=>["&lt;script&gt;alert('Hello');&lt;/script&gt;"] }
    Document.from_params(params) do |doc|
      html = doc.to_html
      assert_match /<input type="hidden" name="NEW_ITEM-LONGTEXT_1:132\[\]" value="&amp;lt;script&amp;gt;alert\(&#39;Hello&#39;\);&amp;lt;\/script&amp;gt;" \/>/, html
    end
  end

private

  def valid_params_complex
    {
      "NEW_ITEM-DESCRIPTION"=>{"1"=>"Visitenkarten"},
      "NEW_ITEM-CURRENCY"=>{"1"=>"EUR"},
      "NEW_ITEM-CUST_FIELD1"=>{"1"=>""},
      "NEW_ITEM-CUST_FIELD2"=>{"1"=>""},
      "NEW_ITEM-CUST_FIELD3"=>{"1"=>""},
      "NEW_ITEM-LEADTIME"=>{"1"=>"10"},
      "NEW_ITEM-MANUFACTCODE"=>{"1"=>""},
      "NEW_ITEM-SERVICE"=>{"1"=>""},
      "NEW_ITEM-EXT_PRODUCT_ID"=>{"1"=>""},
      "NEW_ITEM-CUST_FIELD4"=>{"1"=>""},
      "NEW_ITEM-CONTRACT_ITEM"=>{"1"=>""},
      "NEW_ITEM-EXT_QUOTE_ID"=>{"1"=>""},
      "NEW_ITEM-LONGTEXT_1:132"=>["Standard Visitenkarte deutsch 200 St. "],
      "NEW_ITEM-CUST_FIELD5"=>{"1"=>""},
      "NEW_ITEM-QUANTITY"=>{"1"=>"1"},
      "NEW_ITEM-CONTRACT"=>{"1"=>""},
      "NEW_ITEM-PRICE"=>{"1"=>"16.18"},
      "NEW_ITEM-PRICEUNIT"=>{"1"=>""},
      "NEW_ITEM-VENDOR"=>{"1"=>"9999"},
      "NEW_ITEM-VENDORMAT"=>{"1"=>"1517"},
      "NEW_ITEM-MATGROUP"=>{"1"=>"24-26-09-01"},
      "NEW_ITEM-MANUFACTMAT"=>{"1"=>""},
      "NEW_ITEM-EXT_QUOTE_ITEM"=>{"1"=>""},
      "NEW_ITEM-MATNR"=>{"1"=>""},
      "NEW_ITEM-UNIT"=>{"1"=>"EA"}
    }
  end


end


lib/sapoci/core_ext/blank.rb
class Object #:nodoc:
  unless respond_to?(:blank?)
    def blank?
      respond_to?(:empty?) ? empty? : !self
    end
  end
end

class NilClass #:nodoc:
  unless respond_to?(:blank?)
    def blank?
      true
    end
  end
end

class FalseClass #:nodoc:
  unless respond_to?(:blank?)
    def blank?
      true
    end
  end
end

class TrueClass #:nodoc:
  unless respond_to?(:blank?)
    def blank?
      false
    end
  end
end

class Array #:nodoc:
  unless respond_to?(:blank?)
    alias_method :blank?, :empty?
  end
end

class Hash #:nodoc:
  unless respond_to?(:blank?)
    alias_method :blank?, :empty?
  end
end

class String #:nodoc:
  unless respond_to?(:blank?)
    def blank?
      self !~ /\S/
    end
  end
end

class Numeric #:nodoc:
  unless respond_to?(:blank?)
    def blank?
      false
    end
  end
end


lib/sapoci/item.rb
# encoding: utf-8

require 'rubygems' unless RUBY_VERSION >= '1.9'
require 'nokogiri'
require 'sapoci/core_ext'
require 'bigdecimal'

module SAPOCI
  class Item
    
    attr_reader :index
    attr_accessor :description
    attr_accessor :matnr
    attr_accessor :quantity
    attr_accessor :unit
    attr_accessor :price
    attr_accessor :currency
    attr_accessor :priceunit
    attr_accessor :leadtime
    attr_accessor :longtext
    attr_accessor :vendor
    attr_accessor :vendormat
    attr_accessor :manufactcode
    attr_accessor :manufactmat
    attr_accessor :matgroup
    attr_accessor :service
    attr_accessor :contract
    attr_accessor :contract_item
    attr_accessor :ext_quote_id
    attr_accessor :ext_quote_item
    attr_accessor :ext_product_id
    attr_accessor :attachment
    attr_accessor :attachment_title
    attr_accessor :attachment_purpose
    attr_accessor :ext_schema_type
    attr_accessor :ext_category_id
    attr_accessor :ext_category
    attr_accessor :sld_sys_name
    attr_accessor :cust_field1
    attr_accessor :cust_field2
    attr_accessor :cust_field3
    attr_accessor :cust_field4
    attr_accessor :cust_field5
    
    # Initializes the item.
    def initialize(index)
      @index = index
    end
    
    def quantity
      BigDecimal.new("0#{@quantity.to_s.strip.gsub(/,/,'.')}")
    end

    def quantity_before_type_cast
      @quantity
    end
    
    def price
      BigDecimal.new("0#{@price.to_s.strip.gsub(/,/,'.')}")
    end
    
    def price_before_type_cast
      @price
    end

    def priceunit
      BigDecimal.new("0#{@priceunit.to_s.strip.gsub(/,/,'.')}").nonzero? || 1
    end
    
    def priceunit_before_type_cast
      @priceunit
    end

    def leadtime
      @leadtime.to_i if @leadtime
    end

    def leadtime_before_type_cast
      @leadtime
    end
    
    def service?
      self.service == "X"
    end

    # Returns the item properties as HTML hidden field tags.
    def to_html(options = {})
      html = []
      html << hidden_field_tag("DESCRIPTION",     self.description)     unless self.description.blank?
      html << hidden_field_tag("MATNR",           self.matnr)           unless self.matnr.blank?
      html << hidden_field_tag("QUANTITY",        "%015.3f" % self.quantity)
      html << hidden_field_tag("UNIT",            self.unit)            unless self.unit.blank?
      html << hidden_field_tag("PRICE",           "%015.3f" % self.price)
      html << hidden_field_tag("CURRENCY",        self.currency)        unless self.currency.blank?
      html << hidden_field_tag("PRICEUNIT",       self.priceunit.to_i)  if self.priceunit.to_i > 0
      html << hidden_field_tag("LEADTIME",        "%05d" % self.leadtime) if self.leadtime.to_i > 0
      html << hidden_field_tag("VENDOR",          self.vendor)          unless self.vendor.blank?
      html << hidden_field_tag("VENDORMAT",       self.vendormat)       unless self.vendormat.blank?
      html << hidden_field_tag("MANUFACTCODE",    self.manufactcode)    unless self.manufactcode.blank?
      html << hidden_field_tag("MANUFACTMAT",     self.manufactmat)     unless self.manufactmat.blank?
      html << hidden_field_tag("MATGROUP",        self.matgroup)        unless self.matgroup.blank?
      html << hidden_field_tag("SERVICE",         "X")                  if     self.service?
      html << hidden_field_tag("CONTRACT",        self.contract)        unless self.contract.blank?
      html << hidden_field_tag("CONTRACT_ITEM",   self.contract_item)   unless self.contract_item.blank?
      html << hidden_field_tag("EXT_QUOTE_ID",    self.ext_quote_id)    unless self.ext_quote_id.blank?
      html << hidden_field_tag("EXT_QUOTE_ITEM",  self.ext_quote_item)  unless self.ext_quote_item.blank?
      html << hidden_field_tag("EXT_PRODUCT_ID",  self.ext_product_id)  unless self.ext_product_id.blank?
      html << hidden_field_tag("ATTACHMENT",      self.attachment)      unless self.attachment.blank?
      html << hidden_field_tag("ATTACHMENT_TITLE", self.attachment_title) unless self.attachment_title.blank?
      html << hidden_field_tag("ATTACHMENT_PURPOSE", self.attachment_purpose) unless self.attachment_purpose.blank?
      html << hidden_field_tag("EXT_SCHEMA_TYPE", self.ext_schema_type) unless self.ext_schema_type.blank?
      html << hidden_field_tag("EXT_CATEGORY_ID", self.ext_category_id) unless self.ext_category_id.blank?
      html << hidden_field_tag("EXT_CATEGORY",    self.ext_category)    unless self.ext_category.blank?
      html << hidden_field_tag("SLD_SYS_NAME",    self.sld_sys_name)    unless self.sld_sys_name.blank?
      html << hidden_field_tag("CUST_FIELD1",     self.cust_field1)     unless self.cust_field1.blank?
      html << hidden_field_tag("CUST_FIELD2",     self.cust_field2)     unless self.cust_field2.blank?
      html << hidden_field_tag("CUST_FIELD3",     self.cust_field3)     unless self.cust_field3.blank?
      html << hidden_field_tag("CUST_FIELD4",     self.cust_field4)     unless self.cust_field4.blank?
      html << hidden_field_tag("CUST_FIELD5",     self.cust_field5)     unless self.cust_field5.blank?
      html << "<input type=\"hidden\" name=\"NEW_ITEM-LONGTEXT_#{index + 1}:132[]\" value=\"#{escape_html(self.longtext)}\" />" unless self.longtext.blank?
      html.join
    end
    
  private
    
    def hidden_field_tag(name, value, options = {})
      "<input type=\"hidden\" name=\"NEW_ITEM-#{name}[#{index + 1}]\" value=\"#{escape_html(value)}\" />"
    end
    
    ESCAPE_HTML	=	{ "&" => "&amp;", "<" => "&lt;", ">" => "&gt;", "'" => "&#39;", '"' => "&quot;", }
    ESCAPE_HTML_PATTERN	=	Regexp.union(ESCAPE_HTML.keys)

    # Shamelessly borrowed from Rack::Utils.escape_html(s)
    def escape_html(s)
      s.to_s.gsub(ESCAPE_HTML_PATTERN){|c| ESCAPE_HTML[c] }
    end

  end
end
